stats.go

```diff
@@ -34,3 +36,60 @@ func (a *autographer) addStats(conf configuration) error {
 	log.Infof("Statsd enabled at %s with namespace %s", conf.Statsd.Addr, conf.Statsd.Namespace)
 	return nil
 }
+
+// newStatsdWriter returns a new http.ResponseWriter that sends HTTP response
+// statuses as metrics to statsd. The metric emitted is prefixed with and
+// suffixed with ".response.status.<status code>". The whole metric for
+// "myhandler" will be something like "myhandler.response.status.200". The
+// returned http.ResponseWriter doesn't support the http.Flusher or
+// http.Hijacker type.
+func newStatsdWriter(w http.ResponseWriter, metricPrefix string, stats statsd.ClientInterface) *statsdWriter {
+	return &statsdWriter{ResponseWriter: w, metricPrefix: metricPrefix, stats: stats, headerWritten: new(atomic.Bool)}
+}
+
+var _ http.ResponseWriter = &statsdWriter{}
+
+type statsdWriter struct {
+	http.ResponseWriter
+	metricPrefix string
+	stats        statsd.ClientInterface
+
+	headerWritten *atomic.Bool
+}
+
+func (w *statsdWriter) Write(b []byte) (int, error) {
+	if !w.headerWritten.Load() {
+		w.WriteHeader(http.StatusOK)
+	}
+	return w.ResponseWriter.Write(b)
+}
+
+func (w *statsdWriter) WriteHeader(statusCode int) {
+	if w.headerWritten.CompareAndSwap(false, true) {
+		w.stats.Incr(fmt.Sprintf("%s.response.status.%d", w.metricPrefix, statusCode), nil, 1)
+		w.ResponseWriter.WriteHeader(statusCode)
+	}
+}
+
+// statsMiddleware is an HTTP handler for emitting a statsd metric of request
+// attempts and returns an http.ResponseWriter for recording HTTP response
+// status codes with newStatsdWriter.
+func statsMiddleware(h http.HandlerFunc, handlerName string, stats statsd.ClientInterface) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		stats.Incr(handlerName+".request.attempts", nil, 1)
+		w = newStatsdWriter(w, handlerName, stats)
+		h(w, r)
+	}
+}
+
+// apiStatsMiddleware is an HTTP handler that emits the metrics as described in
+// statsMiddleware, but also increments equivalent metrics as if their
+// handlerName was "api". This is a shortcut to make it easier to get our uptime
+// across the whole over our actual public API.
+func apiStatsMiddleware(h http.HandlerFunc, handlerName string, stats statsd.ClientInterface) http.HandlerFunc {
```

Comment:
It seems unfortunate that we need to duplicate data like this. Could we avoid it if we had a common prefix for all of the things that use this? Eg: `http.api`, to catch both the `sign` and `auths` endpoints?

Reply:
Yeah, that's fair. I'm used to having to do duplicative rollups due to metric renames, etc., but we have a greenfield here. I'll make the change!

Reply:
Done!